Compute horizon quantities at excisions for the zero-expansion BC.

iago-mendes opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

This PR adds support for computing horizon quantities at the excisions. This is needed when creating initial data with the zero-expansion boundary condition, which is now possible to do by passing `--no-negative-expansion-bc` to `spectre bbh generate-id`.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

iago-mendes committed:
Support use of excisions to compute horizon quantities.

nilsvu commented on support/Pipelines/Bbh/FindHorizon.py:
```diff
@@ -12,6 +12,12 @@
+from spectre.Domain import (
+    deserialize_domain,
+    deserialize_functions_of_time,
+    strahlkorper_in_inertial_frame,
```
don't need `strahlkorper_in_inertial_frame`

nilsvu commented on support/Pipelines/Bbh/InitialData.py:
```diff
@@ -147,6 +149,7 @@ def generate_id(
     control: bool = False,
     evolve: bool = False,
     eccentricity_control: bool = False,
+    negative_expansion_bc: bool = True,
```
Add this new argument to the docs

nilsvu commented on support/Pipelines/Bbh/PostprocessId.py:
```diff
@@ -111,22 +116,36 @@ def postprocess_id(
-            obs_id=obs_id,
-            obs_time=0.0,
-            initial_guess=Strahlkorper[Frame.Inertial](
+        horizon_quantities = None
```
don't need this line (Python has more open scopes than C++)

nilsvu commented on support/Pipelines/Bbh/PostprocessId.py:
```diff
@@ -111,22 +116,36 @@ def postprocess_id(
-            obs_time=0.0,
-            initial_guess=Strahlkorper[Frame.Inertial](
+        horizon_quantities = None
+        if not negative_expansion_bc:
```
Generally prefer positive conditions because they are easier to understand (somehow it takes a lot of brain cycles to negate a condition). Just do `if negative_expansion_bc` first.

nilsvu commented on support/Pipelines/Bbh/FindHorizon.py:
```diff
@@ -276,6 +285,122 @@ def find_horizon(
+    for volfile in open_volfiles(h5_files, subfile_name, obs_id):
+        dim = volfile.get_dimension()
+        domain = deserialize_domain[dim](volfile.get_domain(obs_id))
+        functions_of_time = deserialize_functions_of_time(
```
```py
if domain.is_time_dependent():
  # deserialize
else:
  functions_of_time = None
```

nilsvu commented on support/Pipelines/Bbh/FindHorizon.py:
```diff
@@ -276,6 +285,122 @@ def find_horizon(
+            )
+            datfile.append(reduction_data)
+
+    return quantities
```
```py
return excision_inertial, quantities
```

Then this also matches your docs

nilsvu commented on support/Pipelines/Bbh/FindHorizon.py:
```diff
@@ -276,6 +285,122 @@ def find_horizon(
     return strahlkorper, quantities
 
 
+def use_excision_as_horizon(
```
Test this function in `Test_FindHorizon.py`

iago-mendes commented:
I've rebased this repo, fixed the failing ID test, and addressed the comments.

When implementing the horizon finder test, a new error appeared: segmentation fault when `strahlkorper_in_inertial_frame_aligned` is called. I've confirmed that the current code works fine in a BBH ID run, which makes me wonder if there's something peculiar about the test domain that I'm missing. @nilsvu, any ideas?

nilsvu commented on support/Pipelines/Bbh/InitialData.py:
```diff
@@ -209,6 +209,8 @@ def generate_id(
       eccentricity_control: If set to True, an eccentricity reduction script is
         run on the initial data to correct the initial orbital parameters.
+      negative_expansion_bc: If set to True, the excision boundaries are set to
+        be inside of the apparent horizons.
```
Please add something like "this helps to find horizons and start an evolution from the initial data without extrapolation. (default: True)"
Also add this to the CLI help below.
